[JSC] Add numeric hot loop unrolling pattern

hyjorc1 opened the pull request:
#### 9bcfc6eb9f1504539c3be7638d739c0047a2802c
<pre>
[JSC] Add numeric hot loop unrolling pattern
<a href="https://bugs.webkit.org/show_bug.cgi?id=292521">https://bugs.webkit.org/show_bug.cgi?id=292521</a>
<a href="https://rdar.apple.com/150660479">rdar://150660479</a>

Reviewed by Yusuke Suzuki.

This patch introduces a specialized unrolling pattern for &quot;numeric hot loops&quot; — loops
that are dominated by numeric computation and local variable access, and which are not
expected to be inlined due to their size.

Key changes:
- Introduced a new unrolling heuristic for loops where:
  - Only one non-pad block exists in the loop,
  - More than 30% of material nodes are numeric computation nodes,
  - More than 40% are local access nodes,
  - The material node count is between 160 and 225.

Since FTL inlining is governed by the total function size
(`maximumFunctionForCallInlineCandidateBytecodeCostForFTL`, currently 170),
loops with body sizes above 160 are unlikely to be inlined. Therefore, unrolling
them doesn&apos;t harm inlining decisions but may still yield performance benefits.

Additional changes:
- Added `isNumericComputationNode()` and `isLocalAccessNode()` helpers.
- Introduced `MaxNumericHotLoopSize = 225` as the special-case upper bound.
- Refactored `maxAllowedUnrollSize()` to reflect both general and special-case limits.
- Ensured early bailout during loop scanning uses the updated size limit.

This change improves performance on tight computational loops that were previously
skipped due to general unroll thresholds, while avoiding regressions in code size
sensitive scenarios.

Canonical link: <a href="https://commits.webkit.org/294683@main">https://commits.webkit.org/294683@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9ecb0a8aef8cba2ecc52da00f3b998b35a4d10ce

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102694 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22367 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12685 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107859 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Compiled WebKit") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/53335 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22676 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30869 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78106 "Build is in progress. Recent messages:") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/53335 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105700 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17571 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92683 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/58438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/17412 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10715 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52692 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/95369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/87227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10785 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110235 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/101304 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29831 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/21994 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/87092 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30195 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88875 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/86698 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22066 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/31529 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/9255 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/24084 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29758 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35078 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/124937 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29566 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/34680 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32893 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/31128 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -50,6 +50,8 @@ class LoopUnrollingPhase : public Phase {
     using ComparisonFunction = bool (*)(CheckedInt32, CheckedInt32);
     using UpdateFunction = CheckedInt32 (*)(CheckedInt32, CheckedInt32);
 
+    static constexpr uint32_t MaxNumericHotLoopSize = 225;
```
Let's move it to runtime option number to make it tunable easily.

Constellation commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -841,6 +887,67 @@ bool LoopUnrollingPhase::LoopData::isMaterialNode(Graph& graph, Node* node)
+    case JSConstant:
+        if (node->isNumberConstant())
+            return true;
+        FALLTHROUGH;
```
Use `[[fallthrough]]`.

Constellation commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -608,6 +632,23 @@ bool performLoopUnrolling(Graph& graph)
+        return nonPadBlockCount() == 1
+            && ratio(numericComputationCount) > 0.3
+            && ratio(localAccessCount) > 0.4
+            && materialNodeCount > 160
```
Can you add FIXME comment about removing this threshold?

hyjorc1 committed:
[JSC] Add numeric hot loop unrolling pattern
https://bugs.webkit.org/show_bug.cgi?id=292521
rdar://150660479

Reviewed by Yusuke Suzuki.

This patch introduces a specialized unrolling pattern for "numeric hot loops" — loops
that are dominated by numeric computation and local variable access, and which are not
expected to be inlined due to their size.

Key changes:
- Introduced a new unrolling heuristic for loops where:
  - Only one non-pad block exists in the loop,
  - More than 30% of material nodes are numeric computation nodes,
  - More than 40% are local access nodes,
  - The material node count is between 160 and 225.

Since FTL inlining is governed by the total function size
(`maximumFunctionForCallInlineCandidateBytecodeCostForFTL`, currently 170),
loops with body sizes above 160 are unlikely to be inlined. Therefore, unrolling
them doesn't harm inlining decisions but may still yield performance benefits.

Additional changes:
- Added `isNumericComputationNode()` and `isLocalAccessNode()` helpers.
- Introduced `MaxNumericHotLoopSize = 225` as the special-case upper bound.
- Refactored `maxAllowedUnrollSize()` to reflect both general and special-case limits.
- Ensured early bailout during loop scanning uses the updated size limit.

This change improves performance on tight computational loops that were previously
skipped due to general unroll thresholds, while avoiding regressions in code size
sensitive scenarios.

Canonical link: https://commits.webkit.org/294683@main
